### DISCUSSION TITLE
Format.psm1xml required elements for TableColumnHeader

### DISCUSSION THREAD

**Discussion by metablaster** [PowerShell/PowerShell#14608] [Q&A] [answered]
I have defined Format.psm1xml file, relevant part looks like this:

```xml
<TableColumnHeader>
	<Label>ParameterName</Label>
	<Alignment>Left</Alignment>
</TableColumnHeader>

```

Notice that I have omitted the `<width><\width>` element, which will make object size being auto sized instead of some hard coded value.

According to [Format.xsd][link], the `<width>` element is required!

I wanted to confirm this is indeed so and run Trace-Command like this:

```powershell
Trace-Command -Name FormatFileLoading, FormatViewBinding -Expression { Test-SomeCommand }
```

And there was no error at all, formatting works just fine and auto size is automated.

My issue is whether the XSD file documentation is wrong or do I really need to include the `width` element?
Why does it work without it if it's required?

**EDIT:**

Ah I noticed it won't work in all cases, so my second question (in addition to is it required) is, is there any way to make object output auto size? I want to avoid hard coded values. and without using Format-Table ofc.

**EDIT2:**

I figured out for autosize there is an element:
```xml
<TableControl>
	<AutoSize>True</AutoSize>
</TableControl>
```

But is the width still required?

[link]: https://github.com/PowerShell/PowerShell/blob/master/src/Schemas/Format.xsd

**Comment by StartAutomating** [comment, marked as answer]
@metablaster , this gets a little complicated.

Before we get into the answer, I will humbly plug a module I wrote many years ago so that I never had to write .format.ps1xml directly ever again.  It's called [EZOut](https://github.com/StartAutomating/EZOut).  This can be a great saver of time and sanity when writing .format.ps1xml or .types.ps1xml files.

Now, as for how Width works:

Width is not required.  Neither, for that matter, is Autosize.

If you provide a column with, it will ensure that column is _exactly_ N characters.

If you provide AutoSize, all columns without a width will be sized to their largest single entry ( I believe it will truncate the last column at $host.UI.RawUI.BufferSize.Width).

If you do not provide AutoSize, all columns without a width will be be given an equal size. 

Thus if I have 4 columns with no width, and a $host.UI.RawUI.BufferSize.Width of 80, each column would be 20 characters. 
And if I have 4 columns, 2 with no width, and 2 width a width totaling 20, then each column without a fixed width would become 30 characters.

Does this make sense?  Does it help you understand how the <Width></Width> element works?

**Reply by metablaster** [reply]
Thank you, explanation of defaults and their behavior answers my concern.
Safe to assume published xsd comments lack information.
